src/main/java/com/codessquad/qna/service/UserService.java

```diff
@@ -20,31 +24,52 @@ public UserService(UserRepository userRepository) {
         this.userRepository = userRepository;
     }
 
-    public User join(User newUser) {
-        checkRedundancy(newUser);
+    public void join(User newUser) {
+        if (isRedundantUser(newUser)) {
+            throw new JoinFailedException("이미 존재하는 회원입니다.");
+        }
 
-        return userRepository.save(newUser);
+        User savedUser = userRepository.save(newUser);
+        if (!savedUser.equals(newUser)) {
+            throw new JoinFailedException();
+        }
     }
 
-    private void checkRedundancy(User user) {
-        userRepository.findByUserId(user.getUserId())
-                .orElseThrow(IllegalArgumentException::new);
+    public User authenticateUser(String userId, String password) {
+        User user = userRepository.findByUserId(userId)
+                .orElseThrow(() -> new LoginFailedException("존재하지 않는 회원입니다."));
+
+        if (!user.isEqualPassword(password)) {
+            throw new LoginFailedException("비밀번호가 일치하지 않습니다.");
+        }
+
+        return user;
     }
 
-    public List<User> getAllUsers() {
-        return userRepository.findAll();
+    public void updateInfo(User presentUser, User referenceUser, String newPassword) {
+        if (!referenceUser.isEqualPassword(referenceUser.getPassword())) {
+            throw new UnauthorizedAccessException("비밀번호가 일치하지 않습니다.");
+        }
+
+        presentUser.updateUserInfo(referenceUser, newPassword);
+        userRepository.save(presentUser);
     }
 
-    public Optional<User> getOneById(Long id) {
-        return userRepository.findById(id);
+    public Optional<User> getOneByUserId(String userId) {
+        return userRepository.findByUserId(userId);
     }
 
-    public void updateInfo(User presentUser, User referenceUser) {
+    public List<User> getAllUsers() {
+        return userRepository.findAll();
+    }
 
-        referenceUser.setUserId(presentUser.getUserId());
+    public User getOneById(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() -> new NotFoundException("존재하지 않는 회원입니다."));
+    }
 
-        userRepository.delete(presentUser);
-        userRepository.save(referenceUser);
+    private boolean isRedundantUser(User user) {
+        User redundantUser = userRepository.findByUserId(user.getUserId()).orElse(null);
```

Comment:
아래와 같이 바꿀 수 있을 것 같네요.
직접 로직을 구현하기 전에 _미리 구현되어 있는 뭔가가 있지 않나_ 고민하는 습관을 가지는 것이 좋습니다. 

```suggestion
        User redundantUser = userRepository.findByUserId(user.getUserId))).isPresent();

Reply:
그렇군요... Optional에 대해 제대로 공부하지 않고 사용한 부분이 있습니다. 코드에 반영 전에 조금 더 알아보는 습관을 만들어 봐야겠습니다.